gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java

```diff
@@ -267,4 +209,119 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
                 GCUtil.deserializeReference(reference).getName(),
                 gcParams.getDefaultCutOffTimestamp());
   }
+
+  // --- Methods for computing expired content rows ----------
+  /**
+   * convert scala Iterator of reference into scala Iterator of {@link Row}.
+   *
+   * <p>Each reference can produce one or more rows.
+   */
+  private scala.collection.Iterator<Row> getExpiredContentRows(
+      scala.collection.Iterator<String> references,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap,
+      String runId,
+      Timestamp startedAt) {
+    List<Iterator<Row>> iterators = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
+      references.foreach(
+          reference -> {
+            iterators.add(
+                walkAllCommitsInReference(
+                    api,
+                    GCUtil.deserializeReference(reference),
+                    liveContentsBloomFilterMap,
+                    runId,
+                    startedAt));
+            return reference;
+          });
+      // merge the iterators from each task.
+      Iterator<Row> mergedIterator = Iterators.concat(iterators.iterator());
```

Comment:
Please think again - this is *NOT* fixing the issue!

Reply:
The test case was passing, let me understand how it is working.

Reply:
Maybe we can convert the iterator to DF?

Reply:
> Maybe we can convert the iterator to DF?

this is in executor and spark context is not available in executor. map partitions need the iterator<ROW> as output

Reply:
I can close it when the iterator has no more rows per reference. 
For this instead of one API per task, now I have to create one API per reference.

Reply:
This is just moving the problem - not fixing it.

Now you create one API instance per reference, which is just not necessary.
Further, there are cases when the API instance is not closed - for example if not all items have been fetched from the iterator. I.e. you're introducing resource leaks.

Maybe look at how external resources are managed in Spark?

Reply:
> Maybe look at how external resources are managed in Spark?

a) I tried checking it yesterday. Spark suggests broadcasting the resource and cleaning it in driver. As mentioned below in other comment, some things are not serailizable in `NessieHttpClient`. Hence it is a problem.

b) I used this solution in the current implementation
https://stackoverflow.com/questions/36545579/spark-how-to-use-mappartition-and-create-close-connection-per-partition

c) `Now you create one API instance per reference, which is just not necessary.`
because iterators are merged, I cannot check hasNext() of final iterator as it happens in spark code gen. Hence , moved to per reference.

d)`for example if not all items have been fetched from the iterator`, 
this can only happen if task is aborted by the user or executor killed (if executor killed, no need to worry about resource leak on that executor as it is mapped to one JVM). If the task is aborted, I don't have a way to close the api in this situation :(

Reply:
I don't believe that there is no way to properly fix it.
The "not serializable" argument is not valid IMO - otherwise JDBC would not work, CSV file access would not work, etc etc.

Reply:
@ajantha-bhat you should look for ways to share state per executor, maybe: https://able.bio/patrickcording/sharing-objects-in-spark--58x4gbf
will be helpful?
So you should basically create a static `NessieAPI`(that will be initialized per executor)

Reply:
@tomekl007 : Even If I make a static instance at executor, when to close() is till a problem. ( I wouldn't know when all the iterators are consumed, I will only know when each iterator is consumed)

In current code, I am closing when the iterator is consumed fully or any exception in filling iterator. 

But as @snazy said, what if iterator is not consumed fully due to some error (as the final action is `.append()`, iterator is consumed by spark for append() operation, if there is some failure) I don't have a way to close it.

I might have to explore towards serializing the API, but that involves serializing a ton of classes along with it.

Reply:
@snazy : As people are not happy with #3793 (including me). I don't think I have a solution for this other than consuming the iterator and cleaning the resource as mentioned above. But that solution will bloat up the executor.

So, At the moment I don't have any solution.

Reply:
Fixed. Thanks.